File: src/pocketcloud/cloud/provider/CloudMySqlProvider.php
<?php

namespace pocketcloud\cloud\provider;

use pocketcloud\cloud\cache\MaintenanceList;
use pocketcloud\cloud\config\impl\MainConfig;
use pocketcloud\cloud\cache\InGameModule;
use pocketcloud\cloud\group\ServerGroup;
use pocketcloud\cloud\PocketCloud;
use pocketcloud\cloud\provider\database\DatabaseQueries;
use pocketcloud\cloud\provider\migration\MigrationList;
use pocketcloud\cloud\template\Template;
use pocketcloud\cloud\terminal\log\CloudLogger;
use pocketcloud\cloud\util\promise\Promise;
use r3pt1s\mysql\ConnectionPool;
use Throwable;

final class CloudMySqlProvider extends CloudProvider {

    private ?ConnectionPool $connectionPool;

    public function __construct() {
        $this->connectionPool = new ConnectionPool([
            "address" => MainConfig::getInstance()->getMySqlAddress(),
            "user" => MainConfig::getInstance()->getMySqlUser(),
            "password" => MainConfig::getInstance()->getMySqlPassword(),
            "database" => MainConfig::getInstance()->getMySqlDatabase(),
            "port" => MainConfig::getInstance()->getMySqlPort()
        ], 1, PocketCloud::getInstance()->getSleeperHandler(), function (Throwable $throwable): void {
            CloudLogger::get()->error("Something unexpected happened while executing a mysql query...");
            CloudLogger::get()->exception($throwable);
        });

        DatabaseQueries::createTables()->execute(function (): void {
            if (MigrationList::JSON_TO_MYSQL()->checkForMigration()) {
                MigrationList::JSON_TO_MYSQL()->migrate();
            }
        });

        foreach (InGameModule::getAll() as $value) {
            $this->getModuleState($value)->then(fn(bool $v) => InGameModule::setModuleState($value, $v));
        }

        $this->getWhitelist()->then(fn(array $list) => MaintenanceList::sync($list));
    }

    public function addTemplate(Template $template): void {
        DatabaseQueries::addTemplate($template->toArray())->execute();
    }

    public function removeTemplate(Template $template): void {
        DatabaseQueries::removeTemplate($template->getName())->execute();
    }

    public function editTemplate(Template $template, array $newData): void {
        DatabaseQueries::editTemplate($template->getName(), $newData)->execute();
    }

    public function getTemplate(string $template): Promise {
        $promise = new Promise();

        DatabaseQueries::getTemplate($template)
            ->execute(function (?array $result) use($promise): void {
                if (!is_array($result)) {
                    $promise->reject();
                    return;
                }

                if (($template = Template::fromArray($result)) !== null) {
                    $promise->resolve($template);
                } else $promise->reject();
            });

        return $promise;
    }

    public function checkTemplate(string $template): Promise {
        $promise = new Promise();

        DatabaseQueries::checkTemplate($template)
            ->execute(fn(?bool $check) => $promise->resolve($check ?? false));

        return $promise;
    }

    public function getTemplates(): Promise {
        $promise = new Promise();

        DatabaseQueries::getTemplates()
            ->execute(function (?array $result) use($promise): void {
                if (!is_array($result)) {
                    $promise->reject();
                    return;
                }

                $templates = [];
                foreach ($result as $data) {
                    if (($template = Template::fromArray($data)) !== null) {
                        $templates[$template->getName()] = $template;
                    }
                }

                $promise->resolve($templates);
            });

        return $promise;
    }

    public function addServerGroup(ServerGroup $serverGroup): void {
        DatabaseQueries::addServerGroup($serverGroup->toArray(true))->execute();
    }

    public function removeServerGroup(ServerGroup $serverGroup): void {
        DatabaseQueries::removeServerGroup($serverGroup->getName())->execute();
    }

    public function editServerGroup(ServerGroup $serverGroup, array $newData): void {
        if (is_array($newData["templates"])) $newData["templates"] = json_encode($newData["templates"]);
        DatabaseQueries::editServerGroup($serverGroup->getName(), $newData)->execute();
    }

    public function getServerGroup(string $serverGroup): Promise {
        $promise = new Promise();

        DatabaseQueries::getServerGroup($serverGroup)
            ->execute(function (?array $result) use($promise): void {
                if (!is_array($result)) {
                    $promise->reject();
                    return;
                }

                if (($serverGroup = ServerGroup::fromArray($result)) !== null) {
                    $promise->resolve($serverGroup);
                } else $promise->reject();
            });

        return $promise;
    }

    public function checkServerGroup(string $serverGroup): Promise {
        $promise = new Promise();

        DatabaseQueries::checkServerGroup($serverGroup)
            ->execute(fn(?bool $check) => $promise->resolve($check ?? false));

        return $promise;
    }

    public function getServerGroups(): Promise {
        $promise = new Promise();

        DatabaseQueries::getServerGroups()
            ->execute(function (?array $result) use($promise): void {
                if (!is_array($result)) {
                    $promise->reject();
                    return;
                }

                $serverGroups = [];
                foreach ($result as $data) {
                    if (($serverGroup = ServerGroup::fromArray($data)) !== null) {
                        $serverGroups[$serverGroup->getName()] = $serverGroup;
                    }
                }

                $promise->resolve($serverGroups);
            });

        return $promise;
    }

    public function setModuleState(string $module, bool $enabled): void {
        InGameModule::setModuleState($module, $enabled);
        DatabaseQueries::setModuleState($module, $enabled)->execute();
    }

    public function getModuleState(string $module): Promise {
        $promise = new Promise();

        DatabaseQueries::getModuleState($module)
            ->execute(fn(array $result) => $promise->resolve($result["enabled"] == 1));

        return $promise;
    }

    public function enablePlayerNotifications(string $player): void {
        DatabaseQueries::enablePlayerNotifications($player)->execute();
    }

    public function disablePlayerNotifications(string $player): void {
        DatabaseQueries::disablePlayerNotifications($player)->execute();
    }

    public function hasNotificationsEnabled(string $player): Promise {
        $promise = new Promise();

        DatabaseQueries::hasNotificationsEnabled($player)
            ->execute(fn(?bool $enabled) => $promise->resolve($enabled ?? false));

        return $promise;
    }

    public function addToWhitelist(string $player): void {
        DatabaseQueries::addToWhitelist($player)->execute();
        MaintenanceList::add($player);
    }

    public function removeFromWhitelist(string $player): void {
        DatabaseQueries::removeFromWhitelist($player)->execute();
        MaintenanceList::remove($player);
    }

    public function isOnWhitelist(string $player): Promise {
        $promise = new Promise();

        DatabaseQueries::isOnWhitelist($player)
            ->execute(fn(?bool $enabled) => $promise->resolve($enabled ?? false));

        return $promise;
    }

    public function getWhitelist(): Promise {
        $promise = new Promise();

        DatabaseQueries::getWhitelist()
            ->execute(fn(?array $list) => $promise->resolve($list === null ? [] : array_map(fn(array $r) => $r["player"], $list)));

        return $promise;
    }

    public function getConnectionPool(): ?ConnectionPool {
        return $this->connectionPool;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/provider/CloudMySqlProvider.java
package de.pocketcloud.cloud.provider;

import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import de.pocketcloud.api.component.group.IServerGroup;
import de.pocketcloud.api.component.template.ITemplate;
import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.cache.NotificationListCache;
import de.pocketcloud.cloud.cache.WhitelistCache;
import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.provider.database.DatabaseQueries;
import de.pocketcloud.cloud.provider.database.MySqlSettings;
import de.pocketcloud.cloud.template.Template;
import de.pocketcloud.cloud.template.group.ServerGroup;
import de.pocketcloud.common.cache.LocalCache;
import de.pocketcloud.common.concurrent.Promise;
import de.pocketcloud.common.util.FileUtils;

import javax.sql.DataSource;
import java.sql.*;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

public final class CloudMySqlProvider extends CloudProvider {

    private final DataSource connectionPool;

    public CloudMySqlProvider() {
        this.connectionPool = buildConnectionPool(PocketCloud.instance().config().mysqlSettings().asSettings());

        executeAsync(DatabaseQueries.createTables()).thenSuccess(_ -> {
            getWhitelist().thenSuccess(list -> LocalCache.get(WhitelistCache.class).syncIn(list.stream()
                    .collect(Collectors.toMap(Function.identity(), _ -> true))));

            getNotificationList().thenSuccess(list -> LocalCache.get(NotificationListCache.class).syncIn(list.stream()
                    .collect(Collectors.toMap(Function.identity(), _ -> true))));
        });
    }

    @Override
    public Promise<Void> addTemplate(ITemplate template) {
        Map<String, Object> data = template.write();
        return executeAsync(DatabaseQueries.addTemplate(), data.values().toArray());
    }

    @Override
    public Promise<Void> removeTemplate(ITemplate template) {
        return executeAsync(DatabaseQueries.removeTemplate(), template.name());
    }

    @Override
    public Promise<Void> editTemplate(ITemplate template, Map<String, Object> newData) {
        Object[] params = buildUpdateParams(newData, template.name());
        return executeAsync(DatabaseQueries.editTemplate(newData), params);
    }

    @Override
    public Promise<Optional<Template>> getTemplate(String templateName) {
        Promise<Optional<Template>> promise = new Promise<>();
        queryAsync(DatabaseQueries.getTemplate(), templateName).thenSuccess(rows -> {
            if (rows.isEmpty()) {
                promise.resolve(Optional.empty());
                return;
            }

            try {
                promise.resolve(Optional.of(Template.read(rows.getFirst())));
            } catch (Exception e) {
                Promise.rejected(e);
            }
        }).failure(promise::reject);

        return promise;
    }

    @Override
    public Promise<Boolean> checkTemplate(String template) {
        Promise<Boolean> promise = new Promise<>();
        countAsync(DatabaseQueries.checkTemplate(), template)
                .thenSuccess(count -> promise.resolve(count > 0))
                .failure(promise::reject);

        return promise;
    }

    @Override
    public Promise<Map<String, Template>> getTemplates() {
        Promise<Map<String, Template>> promise = new Promise<>();
        queryAsync(DatabaseQueries.getTemplates()).thenSuccess(rows -> {
            Map<String, Template> templates = new HashMap<>();
            rows.forEach(row -> {
                try {
                    Template t = Template.read(row);
                    templates.put(t.name(), t);
                } catch (Exception _) {}
            });

            promise.resolve(templates);
        }).failure(promise::reject);

        return promise;
    }

    @Override
    public Promise<Void> addServerGroup(IServerGroup serverGroup) {
        Map<String, Object> data = serverGroup.write();
        if (data.get("templates") instanceof List) data.put("templates", FileUtils.encodeJson(data.get("templates")));
        return executeAsync(DatabaseQueries.addServerGroup(), data.values().toArray());
    }

    @Override
    public Promise<Void> removeServerGroup(IServerGroup serverGroup) {
        return executeAsync(DatabaseQueries.removeServerGroup(), serverGroup.name());
    }

    @Override
    public Promise<Void> editServerGroup(IServerGroup serverGroup, Map<String, Object> newData) {
        if (newData.get("templates") instanceof List)
            newData.put("templates", FileUtils.encodeJson(newData.get("templates")));
        Object[] params = buildUpdateParams(newData, serverGroup.name());
        return executeAsync(DatabaseQueries.editServerGroup(newData), params);
    }

    @Override
    public Promise<Optional<ServerGroup>> getServerGroup(String serverGroupName) {
        Promise<Optional<ServerGroup>> promise = new Promise<>();
        queryAsync(DatabaseQueries.getServerGroup(), serverGroupName).thenSuccess(rows -> {
            if (rows.isEmpty()) {
                promise.resolve(Optional.empty());
                return;
            }

            try {
                ServerGroup sg = ServerGroup.read(rows.getFirst());
                promise.resolve(Optional.of(sg));
            } catch (Exception e) {
                promise.reject(e);
            }
        }).failure(promise::reject);

        return promise;
    }

    @Override
    public Promise<Boolean> checkServerGroup(String serverGroup) {
        Promise<Boolean> promise = new Promise<>();
        countAsync(DatabaseQueries.checkServerGroup(), serverGroup)
                .thenSuccess(count -> promise.resolve(count > 0))
                .failure(promise::reject);

        return promise;
    }

    @Override
    public Promise<Map<String, ServerGroup>> getServerGroups() {
        Promise<Map<String, ServerGroup>> promise = new Promise<>();
        queryAsync(DatabaseQueries.getServerGroups()).thenSuccess(rows -> {
            Map<String, ServerGroup> serverGroups = new HashMap<>();
            rows.forEach(row -> {
                try {
                    ServerGroup sg = ServerGroup.read(row);
                    serverGroups.put(sg.name(), sg);
                } catch (Exception _) {}
            });
            promise.resolve(serverGroups);
        }).failure(promise::reject);

        return promise;
    }

    @Override
    public Promise<Void> enablePlayerNotifications(String player) {
        LocalCache.get(NotificationListCache.class).add(player, true);
        return executeAsync(DatabaseQueries.enablePlayerNotifications(), player);
    }

    @Override
    public Promise<Void> disablePlayerNotifications(String player) {
        LocalCache.get(NotificationListCache.class).remove(player);
        return executeAsync(DatabaseQueries.disablePlayerNotifications(), player);
    }

    @Override
    public Promise<Boolean> hasNotificationsEnabled(String player) {
        Promise<Boolean> promise = new Promise<>();
        countAsync(DatabaseQueries.hasNotificationsEnabled(), player)
                .thenSuccess(count -> promise.resolve(count > 0))
                .failure(promise::reject);

        return promise;
    }

    @Override
    public Promise<List<String>> getNotificationList() {
        Promise<List<String>> promise = new Promise<>();
        queryAsync(DatabaseQueries.getNotificationList()).
                thenSuccess(rows -> {
                    List<String> list = rows.stream().map(r -> (String) r.get("player")).toList();
                    promise.resolve(list);
                }).failure(promise::reject);

        return promise;
    }

    @Override
    public Promise<Void> addToWhitelist(String player) {
        LocalCache.get(WhitelistCache.class).add(player, true);
        return executeAsync(DatabaseQueries.addToWhitelist(), player);
    }

    @Override
    public Promise<Void> removeFromWhitelist(String player) {
        LocalCache.get(WhitelistCache.class).remove(player);
        return executeAsync(DatabaseQueries.removeFromWhitelist(), player);
    }

    @Override
    public Promise<Boolean> isOnWhitelist(String player) {
        Promise<Boolean> promise = new Promise<>();
        countAsync(DatabaseQueries.isOnWhitelist(), player)
                .thenSuccess(count -> promise.resolve(count > 0))
                .failure(promise::reject);

        return promise;
    }

    @Override
    public Promise<List<String>> getWhitelist() {
        Promise<List<String>> promise = new Promise<>();
        queryAsync(DatabaseQueries.getWhitelist()).thenSuccess(rows -> {
            List<String> list = rows.stream().map(r -> (String) r.get("player")).toList();
            promise.resolve(list);
        }).failure(promise::reject);

        return promise;
    }

    private Promise<Void> executeAsync(String sql, Object... params) {
        return Promise.runAsync(() -> {
            try (Connection conn = connectionPool.getConnection(); PreparedStatement ps = conn.prepareStatement(sql)) {
                bindParams(ps, params);
                ps.executeUpdate();
            } catch (SQLException e) {
                CloudLogger.get().exception("MySQL execute rejecteded: {}", e, sql);
                throw new RuntimeException(e);
            }
        });
    }

    private Promise<List<Map<String, Object>>> queryAsync(String sql, Object... params) {
        return Promise.supplyAsync(() -> {
            List<Map<String, Object>> rows = new ArrayList<>();
            try (Connection conn = connectionPool.getConnection();
                 PreparedStatement ps = conn.prepareStatement(sql)) {
                bindParams(ps, params);
                try (ResultSet rs = ps.executeQuery()) {
                    ResultSetMetaData meta = rs.getMetaData();
                    int cols = meta.getColumnCount();
                    while (rs.next()) {
                        Map<String, Object> row = new HashMap<>();
                        for (int i = 1; i <= cols; i++) {
                            row.put(meta.getColumnLabel(i), rs.getObject(i));
                        }
                        rows.add(row);
                    }
                }
            } catch (SQLException e) {
                CloudLogger.get().exception("MySQL query rejecteded: {}", e, sql);
                throw new RuntimeException(e);
            }
            return rows;
        });
    }

    private Promise<Integer> countAsync(String sql, Object... params) {
        return queryAsync(sql, params).thenApply(rows -> {
            if (rows.isEmpty()) return 0;
            Object v = rows.getFirst().values().iterator().next();
            return v instanceof Number ? ((Number) v).intValue() : 0;
        });
    }

    private static void bindParams(PreparedStatement ps, Object[] params) throws SQLException {
        for (int i = 0; i < params.length; i++) {
            ps.setObject(i + 1, params[i]);
        }
    }

    private static Object[] buildUpdateParams(Map<String, Object> data, String pkValue) {
        List<Object> params = new ArrayList<>(data.values());
        params.add(pkValue);
        return params.toArray();
    }

    private static DataSource buildConnectionPool(MySqlSettings mysqlSettings) {
        HikariConfig config = new HikariConfig();

        config.setJdbcUrl("jdbc:mysql://" + mysqlSettings.address() + ":" + mysqlSettings.port() + "/" + mysqlSettings.database());
        config.setUsername(mysqlSettings.user());
        config.setPassword(mysqlSettings.password());

        return new HikariDataSource(config);
    }
}